File: mobile/lib/core/api/feature_disabled.dart
import 'package:dio/dio.dart';

/// WIRE FORMAT — phải khớp `FEATURE_DISABLED_ERROR` trong
/// `backend/src/feature-flags/guards/feature-flag.guard.ts`.
///
/// Bản đã cài trên máy người dùng mang theo bản sao của chuỗi này. Đổi tên ở
/// backend là mọi APK cũ lặng lẽ quay về hiện 404 thô — không có deploy nào cứu
/// được, vì lỗi nằm trong ứng dụng đã cài.
const String kFeatureDisabledError = 'FEATURE_DISABLED';

/// Một phân hệ bị tắt, đọc từ thân lỗi 404 của backend.
class FeatureDisabled {
  const FeatureDisabled({required this.feature, required this.message});

  final String feature;
  final String message;
}

/// Đọc `DioException` xem có phải "tính năng đang tắt" không.
///
/// Backend giữ mã 404 có chủ ý — mã riêng sẽ giúp người ngoài dò được phân hệ
/// nào đang tắt — nên chỉ có THÂN lỗi phân biệt được "tính năng tắt" với "không
/// tìm thấy bản ghi". Không đọc thân thì màn hình chi tiết hiện
/// `Lỗi: DioException ... 404` cho một cán bộ không làm gì sai.
///
/// Nhận BA hình dạng, và hình dạng thứ ba mới là hình dạng máy chủ THẬT trả về:
///
///  1. phẳng — `{ error: 'FEATURE_DISABLED', feature, message }`, đúng thứ
///     `FeatureFlagGuard` ném ra;
///  2. lồng dưới `message` — một dạng Nest có thể sinh;
///  3. **bọc bởi `GlobalExceptionFilter`** — `{ error: { code:
///     'FEATURE_DISABLED', message } }`. Filter bọc MỌI `HttpException`, nên mã
///     phẳng ở (1) trở thành `error.code`.
///
/// Trước đây chỉ nhận (1) và (2), nên app **chưa bao giờ** nhận diện được tính
/// năng bị tắt: nó luôn hiện lỗi chung. Không ai thấy vì gate cờ chưa bao giờ
/// chạy (ADR-0018) — sửa gate xong mới lộ ra hợp đồng hai đầu không khớp.
FeatureDisabled? readFeatureDisabled(DioException err) {
  if (err.response?.statusCode != 404) return null;
  final data = err.response?.data;
  if (data is! Map) return null;

  Map? holder;
  if (data['error'] == kFeatureDisabledError) {
    holder = data;
  } else {
    final nested = data['message'];
    if (nested is Map && nested['error'] == kFeatureDisabledError) {
      holder = nested;
    } else {
      // Dạng bọc bởi `GlobalExceptionFilter` — dạng thật trên dây.
      final wrapped = data['error'];
      if (wrapped is Map && wrapped['code'] == kFeatureDisabledError) {
        holder = wrapped;
      }
    }
  }
  if (holder == null) return null;

  return FeatureDisabled(
    feature: holder['feature'] as String? ?? '',
    message: holder['message'] as String? ?? 'Tính năng hiện đang tắt',
  );
}

File: frontend/src/lib/feature-disabled.ts
/**
 * Telling "this feature is switched off" apart from "that record is gone".
 *
 * `FeatureFlagGuard` answers a gated route with 404 — deliberately, so a
 * disabled feature does not advertise itself with a distinct status code — but
 * puts `error: 'FEATURE_DISABLED'` in the body. Without reading that body every
 * screen showed its own generic not-found message, which reads to an officer
 * like their case file vanished rather than like an administrator flipped a
 * switch.
 */

/**
 * WIRE FORMAT — must match `FEATURE_DISABLED_ERROR` in
 * `backend/src/feature-flags/guards/feature-flag.guard.ts`. The mobile client
 * carries its own copy, and an installed APK cannot be corrected by a redeploy.
 */
export const FEATURE_DISABLED_ERROR = 'FEATURE_DISABLED';

/** Name of the window event the API interceptor broadcasts. */
export const FEATURE_DISABLED_EVENT = 'pc02:feature-disabled';

export interface FeatureDisabledDetail {
  /** Flag key, e.g. `"cases"`. */
  feature: string;
  /** Vietnamese sentence from the server, ready to display. */
  message: string;
}

interface MaybeAxiosError {
  response?: {
    status?: number;
    data?: {
      error?: unknown;
      feature?: unknown;
      message?: unknown;
      // Nest wraps a thrown object literal under `message` on some paths, so
      // accept both shapes rather than depending on which one the framework
      // happens to produce.
      [key: string]: unknown;
    };
  };
}

function body(error: unknown): Record<string, unknown> | undefined {
  const data = (error as MaybeAxiosError)?.response?.data;
  if (!data || typeof data !== 'object') return undefined;
  const nested = (data as { message?: unknown }).message;
  if (nested && typeof nested === 'object') {
    return nested as Record<string, unknown>;
  }
  return data as Record<string, unknown>;
}

/**
 * Mã nằm ở đâu trong thân lỗi — phải chấp nhận HAI dạng.
 *
 * `FeatureFlagGuard` ném `{ statusCode, error: 'FEATURE_DISABLED', feature,
 * message }` (dạng phẳng). Nhưng `GlobalExceptionFilter` BỌC LẠI mọi
 * `HttpException` thành `{ success: false, error: { code, message, details } }`
 * — mã phẳng trở thành `error.code`.
 *
 * Client chỉ đọc dạng phẳng nên **chưa bao giờ nhận diện được** tính năng bị
 * tắt: nó luôn rơi vào nhánh lỗi chung. Không ai thấy vì gate chưa bao giờ chạy
 * (ADR-0018) — sửa gate xong mới lộ ra hợp đồng hai đầu không khớp nhau.
 *
 * Giữ cả hai nhánh chứ không chỉ đổi sang dạng bọc: dạng phẳng vẫn tới được
 * client nếu có đường nào bỏ qua filter.
 */
function featureDisabledCode(b: Record<string, unknown> | undefined): unknown {
  if (!b) return undefined;
  if (b.error === FEATURE_DISABLED_ERROR) return b.error;
  const nested = b.error;
  if (nested && typeof nested === 'object') {
    return (nested as Record<string, unknown>).code;
  }
  return undefined;
}

export function isFeatureDisabledError(error: unknown): boolean {
  const status = (error as MaybeAxiosError)?.response?.status;
  if (status !== 404) return false;
  return featureDisabledCode(body(error)) === FEATURE_DISABLED_ERROR;
}

export function extractFeatureDisabled(
  error: unknown,
): FeatureDisabledDetail {
  const b = body(error) ?? {};
  // Dạng bọc đặt `message` bên trong `error`; dạng phẳng đặt ở ngoài.
  const wrapped =
    b.error && typeof b.error === 'object'
      ? (b.error as Record<string, unknown>)
      : {};
  const feature =
    typeof b.feature === 'string'
      ? b.feature
      : typeof wrapped.feature === 'string'
        ? wrapped.feature
        : '';
  const rawMessage =
    typeof b.message === 'string' && b.message.length > 0
      ? b.message
      : typeof wrapped.message === 'string'
        ? wrapped.message
        : '';
  const message = rawMessage.length > 0 ? rawMessage : 'Tính năng này hiện đang tắt.';
  return { feature, message };
}
